### DISCUSSION TITLE
Field meta data

### DISCUSSION THREAD

**Discussion by wdfinch** [react-hook-form/react-hook-form#3975] [Ideas]
A useful feature would be the ability to add metadata to an input. In order the do this now, I make the input and object and then write a custom `value` property and `onChange` method to parse through and pull out the part the gets rendered. 

So my text input might look like `{text: 'this is what's is rendered', metaData: {}}` and then I need to override the `value` and `onChange` on the `<Controller/>` to extract/update just the text part. 

It would be cool if there was a way to do this that was easier and more native. So for example a method called `setMeta(fieldName: string, meta: object)`.

**Comment by bluebill1049** [comment]
what do you mean by `metaData`? data associated with an actual input?

**Reply by wdfinch** [reply]
Yes, so an example might be this: 

Let's say I get a list of items from a server and each item is displayed as a few inputs in a field array. One thing that would also be helpful, is to associate the `id` (`id` from the server) of the item with the inputs. So that way on submit, I could compare what changed. 

I could use `setValue` to add the `id` in, but the issue is I can't set a value unless the input is rendered on the dom. The only ways to do this are to have a hidden input which renders the `id`, store the `id` in a separate list and keep track of which index it's associated with in the field array, or make one the input have the `id` and use custom `value`/`onChange` methods to set/extract it. 

I guess the easiest solution would be to allow the user to set values without requiring that the value be rendered to the dom, but this might be covered by V7. A cleaner approach would be to set meta data since values should be reserved for what get's rendered. 

Does that make sense @bluebill1049?

**Reply by Noitidart** [reply]
I had this idea and am using successfully - https://github.com/react-hook-form/react-hook-form/discussions/8161

**Reply by ramseylove** [reply]
@wdfinch Did you happen to figure out a way to implement this or was anything added to v7 to make this easier for you?

**Comment by stijnvanderlaan** [comment]
This would be very useful, because adding an controller for an none existing field sets the state of the form dirty, and using useWatch doesn't update when the values changes: https://github.com/react-hook-form/react-hook-form/discussions/8721

I try to mark a field as conflicted with the server version but this is very hard to do now.

**Comment by bombillazo** [comment]
This could be done with `rhf-plus` using the form metadata, you can add your own custom data and track it per field using the field as the metadata key

https://github.com/bombillazo/rhf-plus/blob/master/docs/form-metadata.md
